File: app/src/main/java/com/djacoronel/sleeklauncher/settings/BackgroundSettingsActivity.java
package com.djacoronel.sleeklauncher.settings;

import android.app.Activity;
import android.app.WallpaperManager;
import android.content.Context;
import android.content.SharedPreferences;
import android.graphics.Bitmap;
import android.graphics.Color;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.ImageView;
import android.widget.SeekBar;
import android.widget.TabHost;
import android.widget.TextView;

import com.djacoronel.sleeklauncher.R;

public class BackgroundSettingsActivity extends Activity {
    ImageView preview;
    int argb[] = {0, 0, 0, 0};
    int blur = 0;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_background_settings);

        setupTabs();
        setupApplyButton();

        SeekBar alphaBar = (SeekBar) findViewById(R.id.transparencyBar);
        SeekBar redBar = (SeekBar) findViewById(R.id.redBar);
        SeekBar greenBar = (SeekBar) findViewById(R.id.greenBar);
        SeekBar blueBar = (SeekBar) findViewById(R.id.blueBar);
        SeekBar blurBar = (SeekBar) findViewById(R.id.blurBar);

        setupBackgroundPreview();
        setupArgbSeekBar(alphaBar, 0);
        setupArgbSeekBar(redBar, 1);
        setupArgbSeekBar(greenBar, 2);
        setupArgbSeekBar(blueBar, 3);
        setupBlurSeekBar(blurBar);
    }

    public void setupTabs() {
        TabHost host = (TabHost) findViewById(R.id.tabHost);
        host.setup();

        //Tab 1
        TabHost.TabSpec spec = host.newTabSpec("Tab One");
        spec.setContent(R.id.tab1);
        spec.setIndicator("Color");
        host.addTab(spec);

        //Tab 2
        spec = host.newTabSpec("Tab Two");
        spec.setContent(R.id.tab2);
        spec.setIndicator("Blur");
        host.addTab(spec);
    }

    void setupApplyButton(){
        TextView apply = (TextView) findViewById(R.id.apply);
        apply.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                savePref();
                finish();
            }
        });
    }

    public void setupBackgroundPreview() {
        preview = (ImageView) findViewById(R.id.preview);
        preview.setScaleType(ImageView.ScaleType.CENTER_CROP);
        Drawable wallpaperDrawable = WallpaperManager.getInstance(this).getDrawable();
        Bitmap wallpaperBitmap = ((BitmapDrawable) wallpaperDrawable).getBitmap();
        preview.setImageDrawable(new BitmapDrawable(getResources(), wallpaperBitmap));
    }

    public void setupArgbSeekBar(SeekBar seekBar, final int argbIndex) {
        seekBar.setMax(255);
        seekBar.setProgress(1);
        seekBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                argb[argbIndex] = progress;
                preview.setColorFilter(Color.argb(argb[0], argb[1], argb[2], argb[3]));
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {

            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {

            }
        });
    }

    public void setupBlurSeekBar(SeekBar blurBar) {
        blurBar.setMax(75);
        blurBar.setProgress(1);
        blurBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                applyBlur(progress);
                blur = progress;
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {

            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {
            }
        });
    }

    public void applyBlur(int progress) {
        Drawable wallpaperDrawable = WallpaperManager.getInstance(this).getDrawable();
        Bitmap blurredBitmap = ((BitmapDrawable) wallpaperDrawable).getBitmap();

        int numberOfFullBlur = progress / 25;
        int remainingBlur = progress % 25;

        for (int i = 0; i < numberOfFullBlur; i++) {
            blurredBitmap = BlurBuilder.blur(this, blurredBitmap, 25);
        }

        if (remainingBlur != 0) {
            blurredBitmap = BlurBuilder.blur(this, blurredBitmap, remainingBlur);
        }

        preview.setImageDrawable(new BitmapDrawable(getResources(), blurredBitmap));
    }

    public void savePref() {
        SharedPreferences sharedPref = PreferenceManager.getDefaultSharedPreferences(this);
        SharedPreferences.Editor editor = sharedPref.edit();

        String bgPref = argb[0] + "," + argb[1] + "," + argb[2] + "," + argb[3] + "," + blur;

        editor.putString("bgPref", bgPref);
        editor.apply();
    }
}

File: app/src/main/java/com/djacoronel/sleeklauncher/settings/BackgroundSettingsActivity.kt
package com.djacoronel.sleeklauncher.settings

import android.app.Activity
import android.app.WallpaperManager
import android.content.SharedPreferences
import android.graphics.Color
import android.graphics.drawable.BitmapDrawable
import android.os.Bundle
import android.view.View
import android.widget.ImageView
import android.widget.SeekBar
import android.widget.TabHost
import com.djacoronel.sleeklauncher.R
import dagger.android.AndroidInjection
import kotlinx.android.synthetic.main.activity_background_settings.*
import kotlinx.android.synthetic.main.blur_transparency_layout.*
import kotlinx.android.synthetic.main.color_picker_layout.*
import javax.inject.Inject

class BackgroundSettingsActivity : Activity() {

    @Inject
    lateinit var sharedPref: SharedPreferences

    private var argb = intArrayOf(0, 0, 0, 0)
    private var blur = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        AndroidInjection.inject(this)
        setContentView(R.layout.activity_background_settings)

        setupTabs()
        setupApplyButton()

        setupBackgroundPreview()
        setupArgbSeekBar(transparencyBar, 0)
        setupArgbSeekBar(redBar, 1)
        setupArgbSeekBar(greenBar, 2)
        setupArgbSeekBar(blueBar, 3)
        setupBlurSeekBar(blurBar)

        loadCurrentPrefs()
    }

    private fun setupTabs() {
        tabHost.setup()

        //Tab 1
        var spec: TabHost.TabSpec = tabHost.newTabSpec("Tab One")
        spec.setContent(R.id.tab1)
        spec.setIndicator("Color")
        tabHost.addTab(spec)

        //Tab 2
        spec = tabHost.newTabSpec("Tab Two")
        spec.setContent(R.id.tab2)
        spec.setIndicator("Blur")
        tabHost.addTab(spec)
    }

    private fun setupApplyButton() {
        apply.setOnClickListener {
            savePref()
            finish()
        }
    }

    private fun setupBackgroundPreview() {
        preview.scaleType = ImageView.ScaleType.CENTER_CROP
        val wallpaperDrawable = WallpaperManager.getInstance(this).drawable
        val wallpaperBitmap = (wallpaperDrawable as BitmapDrawable).bitmap
        preview.setImageDrawable(BitmapDrawable(resources, wallpaperBitmap))
    }

    private fun setupArgbSeekBar(seekBar: SeekBar, argbIndex: Int) {
        seekBar.max = 255
        seekBar.progress = 1
        seekBar.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                argb[argbIndex] = progress
                preview.setColorFilter(Color.argb(argb[0], argb[1], argb[2], argb[3]))
            }

            override fun onStartTrackingTouch(seekBar: SeekBar) {}
            override fun onStopTrackingTouch(seekBar: SeekBar) {}
        })
    }

    private fun setupBlurSeekBar(blurBar: SeekBar) {
        blurBar.max = 50
        blurBar.progress = 1
        blurBar.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                applyBlur(progress)
                blur = progress
            }

            override fun onStartTrackingTouch(seekBar: SeekBar) {}
            override fun onStopTrackingTouch(seekBar: SeekBar) {}
        })
    }

    private fun applyBlur(progress: Int) {
        val wallpaperDrawable = WallpaperManager.getInstance(this).drawable
        var blurredBitmap = (wallpaperDrawable as BitmapDrawable).bitmap

        val numberOfFullBlur = progress / 25
        val remainingBlur = progress % 25

        for (i in 0 until numberOfFullBlur)
            blurredBitmap = BlurBuilder().blur(this, blurredBitmap, 25f)

        if (remainingBlur != 0)
            blurredBitmap = BlurBuilder().blur(this, blurredBitmap, remainingBlur.toFloat())

        preview.setImageDrawable(BitmapDrawable(resources, blurredBitmap))
    }

    private fun loadCurrentPrefs() {
        val storedBgPref = sharedPref.getString("bgPref", "0,0,0,0,0")
        val splitStoredBgPref = storedBgPref.split(",")

        transparencyBar.progress = splitStoredBgPref[0].toInt()
        redBar.progress = splitStoredBgPref[1].toInt()
        greenBar.progress = splitStoredBgPref[2].toInt()
        blueBar.progress = splitStoredBgPref[3].toInt()
        blurBar.progress = splitStoredBgPref[4].toInt()
    }

    private fun savePref() {
        val editor = sharedPref.edit()
        val bgPref = "${argb[0]},${argb[1]},${argb[2]},${argb[3]},$blur"
        editor.putString("bgPref", bgPref)
        editor.apply()
    }
}